### PR TITLE
[BUGFIX] Fix s3 path suffix bug on windows

### LINKED ISSUE #4022: S3 prefixes get broken when running on Windows
**Describe the bug**
On Windows when referencing an S3 data source, the prefix is suffixed with ``\`` instead of ``/``, resulting in query always failing.

**To Reproduce**
Create a ``InferredAssetS3DataConnector`` data connector, and pass any valid prefix (probably needs to be non-empty). Then, attempt to load a file from that connector.

**Expected behavior**
The file gets loaded properly.

**Error**
```
...\lib\site-packages\great_expectations\datasource\data_connector\util.py in list_s3_keys(s3, query_options, iterator_dict, recursive)
    452 
    453     if not any(key in s3_objects_info for key in ["Contents", "CommonPrefixes"]):
--> 454         raise ValueError("S3 query may not have been configured correctly.")
    455 
    456     if "Contents" in s3_objects_info:

ValueError: S3 query may not have been configured correctly.
```

**Fix**
In ``inferred_asset_s3_data_connector.py``, line 82 is:

```
        self._prefix = FilePathDataConnector.sanitize_prefix(prefix)
```

This ensures that the file path is always a folder, by appending a slash to the end. However, on Windows, the slash is ``\``, and thus you end up with a prefix like ``path/to/some/folder\``, which results in an S3 query with no results, hence the above stacktrace.

The fix is to ensure that only the correct slash can be appended to the prefix, regardless of the host OS. For example:

```
        self._prefix = prefix.rstrip('/') + '/'
```

or something to that effect.

**Environment (please complete the following information):**
 - Operating System: Windows
 - Great Expectations Version: 0.14.1

### PR DESCRIPTION
**Pull request by scnerd**
Resolves #4022 

Changes proposed in this pull request:
- Updates ``InferredAssetS3DataConnector`` to use a custom file path sanitizer that does not rely on the host OS, but assumes (according to S3 convention) a ``/`` separator.

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by scnerd** [comment]
@cla-bot check

**Comment by scnerd** [comment]
@cla-bot check

File: great_expectations/datasource/data_connector/inferred_asset_s3_data_connector.py [hunk 1]
```diff
@@ -93,6 +93,22 @@ def __init__(
                 "Unable to load boto3 (it is required for InferredAssetS3DataConnector)."
             )
 
+    @staticmethod
+    def sanitize_prefix(text: str) -> str:
```
Comment (talagluck) [inline great_expectations/datasource/data_connector/inferred_asset_s3_data_connector.py]: Could we change the name of this to `sanitize_prefix_for_s3` unless this will be used for anything else?

Additionally, would you please add this to the `ConfiguredAssetS3DataConnector`?

**Review by talagluck — COMMENTED** [review]
Hi @scnerd - thanks so much for your contribution! In addition to the feedback within, it looks like tests are failing here - would you please take a look and resolve?

**Comment by kyleaton** [comment]
@cla-bot check

**Comment by scnerd** [comment]
@cla-bot check

**Comment by scnerd** [comment]
@talagluck Looks like all tests are passing now, could you review again? Thanks

File: great_expectations/datasource/data_connector/configured_asset_s3_data_connector.py [hunk 2]
```diff
@@ -96,6 +93,26 @@ def __init__(
...
+        if not text:
+            return text
+
+        path_parts = text.split("/")
+        if not path_parts:  # Empty prefix
+            return text
+        elif "." in path_parts[-1]:  # File, not folder
+            return text
+        else:  # Folder, should have trailing /
+            return text.rstrip("/") + "/"
```
Comment (AFineDayFor) [inline great_expectations/datasource/data_connector/configured_asset_s3_data_connector.py:114]: NIT: Is it okay to assume that every base name should be manipulated like this, and would it break an existing workflow.
A directory name could have a fullstops in this. how would you go about addressing it if this were the case?

This appears to be something that can happen, and I believe it's possible for a dot to surface in a directory name, what's a suggested workaround for cases like that?

Comment (AFineDayFor) [reply great_expectations/datasource/data_connector/configured_asset_s3_data_connector.py:114]: We'd generally lean towards similar logic as [here](https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/data_context/store/tuple_store_backend.py#L478-L491)

**Review by AFineDayFor — CHANGES_REQUESTED** [review]

File: great_expectations/datasource/data_connector/configured_asset_s3_data_connector.py (hunk 2 above)
Comment (scnerd) [reply great_expectations/datasource/data_connector/configured_asset_s3_data_connector.py:114]: @AFineDayFor My target was to provide identical behavior as the original sanitizer ``FilePathDataConnector.sanitize_prefix``, with the exception being that it did not rely on system path separators. I also avoided merely wrapping the original sanitizer with ``.replace('\\', '/')`` since both slashes are allowed in S3 paths.

I've added a new test that ensures that the new sanitizer results in exactly the same paths as the original sanitizer (following all the test cases for the original sanitizer, as well as a few additional cases). There's a little cleanup in that test case to allow it to pass on Windows, but it's worth noting that, without that cleanup, it wouldn't work on Windows, and that cleanup (as mentioned above) is not appropriate for production use. Please check the cases I list in the new test and let me know if there are any edge cases you're still worried about.

File: tests/datasource/data_connector/test_configured_asset_s3_data_connector.py [hunk 3]
```diff
@@ -1117,3 +1121,31 @@ def test_example_with_explicit_data_asset_names():
...
+    check_sameness("foo/", "foo/")
+    check_sameness("bar", "bar/")
+    check_sameness("baz.txt", "baz.txt")
+    check_sameness("a/b/c/baz.txt", "a/b/c/baz.txt")
+
+    # A couple additional checks
+    check_sameness("a/b/c", "a/b/c/")
+    check_sameness("a.x/b/c", "a.x/b/c/")
+    check_sameness("path/to/folder.something/", "path/to/folder.something/")
+    check_sameness("path/to/folder.something", "path/to/folder.something")
```
Comment (scnerd) [inline tests/datasource/data_connector/test_configured_asset_s3_data_connector.py:1151]: @AFineDayFor Here are the equality checks performed to ensure the new sanitizer does not break user code.

**Review by AFineDayFor — APPROVED** [review]

**Review by AFineDayFor — APPROVED** [review]

**Comment by AFineDayFor** [comment]
Thank you much @scnerd.
